tests/e2e/k8s/suite.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2023 The ClusterLink Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/stretchr/testify/suite"
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/clusterlink-net/clusterlink/pkg/bootstrap/platform"
+	"github.com/clusterlink-net/clusterlink/tests/e2e/k8s/util"
+)
+
+const (
+	clusterCount     = 3
+	exportedLogsPath = "/tmp/clusterlink-k8s-tests"
+)
+
+var images = [...]string{"cl-controlplane", "cl-dataplane", "cl-go-dataplane"}
+
+var iperfService = util.Service{
+	Name:      "iperf3-server",
+	Namespace: v1.NamespaceDefault,
+	Port:      80,
+}
+
+var httpEchoService = util.Service{
+	Name:      "http-echo",
+	Namespace: v1.NamespaceDefault,
+	Port:      80,
+}
+
+func httpEchoPodAndService(echoValue string) *util.PodAndService {
+	return &util.PodAndService{
+		Service:    httpEchoService,
+		Image:      "hashicorp/http-echo",
+		Args:       []string{"-listen=:8080", "-text=" + echoValue},
+		TargetPort: 8080,
+	}
+}
+
+// TestSuite is a suite for e2e testing on k8s clusters.
+type TestSuite struct {
+	suite.Suite
+	fabric   *util.Fabric
+	clusters []*util.KindCluster
+}
+
+// SetupSuite creates the k8s kind clusters and the clusterlink certificates.
+func (s *TestSuite) SetupSuite() {
+	fabric, err := util.NewFabric()
+	if err != nil {
+		s.T().Fatal(err)
+	}
+
+	// create clusters and fabric
+	s.fabric = fabric
+	s.clusters = make([]*util.KindCluster, clusterCount)
+	for i := 0; i < clusterCount; i++ {
+		s.clusters[i] = util.NewKindCluster(fmt.Sprintf("peer%d", i))
+		fabric.CreatePeer(s.clusters[i])
+
+		s.clusters[i].Start()
+		for _, image := range images {
+			s.clusters[i].LoadImage(image)
+		}
+	}
+
+	// prepare logs directory
+	if err := os.RemoveAll(exportedLogsPath); err != nil {
+		s.T().Fatal(fmt.Errorf("cannot cleanup logs directory: %w", err))
+	}
+	if err := os.MkdirAll(exportedLogsPath, 0755); err != nil {
+		s.T().Fatal(fmt.Errorf("cannot create logs directory: %w", err))
+	}
+
+	// wait for clusters
+	for i := 0; i < clusterCount; i++ {
+		if err := s.clusters[i].Wait(); err != nil {
+			s.T().Fatal(err)
+		}
+
+		// create http-echo service which echos the cluster name
```

Comment:
Since these services are expected to be used by many tests, I don't want to waste time in re-launching them for each test.
This is why I prefer to create them at the suite setup.
It should not matter to the tests that these services pre-existed as they are stateless.

Reply:
Understood.
My suggestion is to make these configurable when creating the test suite and not hard coded in the test suite. If I come up with tests for k8s based management (ie CRDs), I will also get these services even though I won't need or want them.

Reply:
You assume there will be multiple suite instances using this code? Why?

Reply:
Once we implement a k8s operator, we could change the code in `fabric.go` to use that.
The rest should be re-used.